File: src/main/java/com/sweetraingarden/gplugin/moduledot/ModuleDotPlugin.java
package com.sweetraingarden.gplugin.moduledot;

import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.Task;
import org.jgrapht.Graph;
import org.jgrapht.graph.DefaultDirectedGraph;
import org.jgrapht.graph.DefaultEdge;
import org.jgrapht.nio.DefaultAttribute;
import org.jgrapht.nio.dot.DOTExporter;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.LinkedHashMap;
import java.util.Map;

public class ModuleDotPlugin implements Plugin<Project> {
    @Override
    public void apply(Project project) {
        project.getTasks().register("generateModuleDotGraph", task -> {
            task.setGroup("documentation");
            task.setDescription("Generates a DOT graph of module dependencies");
            
            task.doLast(t -> {
                try {
                    generateDotGraph(project);
                } catch (IOException e) {
                    throw new RuntimeException("Failed to generate module dependency graph", e);
                }
            });
        });
    }

    public void generateDotGraph(Project project) throws IOException {
        Graph<String, DefaultEdge> graph = new DefaultDirectedGraph<>(DefaultEdge.class);
        
        // Add all projects to the graph
        project.getAllprojects().forEach(p -> graph.addVertex(p.getName()));
        
        // Add dependencies between projects
        project.getAllprojects().forEach(p -> {
            p.getConfigurations().forEach(config -> {
                config.getDependencies().forEach(dep -> {
                    if (dep instanceof org.gradle.api.artifacts.ProjectDependency) {
                        org.gradle.api.artifacts.ProjectDependency projectDep = 
                            (org.gradle.api.artifacts.ProjectDependency) dep;
                        graph.addEdge(p.getName(), projectDep.getDependencyProject().getName());
                    }
                });
            });
        });

        // Create DOT exporter
        DOTExporter<String, DefaultEdge> exporter = new DOTExporter<>();
        exporter.setVertexAttributeProvider((v) -> {
            Map<String, org.jgrapht.nio.Attribute> map = new LinkedHashMap<>();
            map.put("label", DefaultAttribute.createAttribute(v));
            map.put("shape", DefaultAttribute.createAttribute("box"));
            return map;
        });

        // Export to file
        File outputFile = project.getLayout().getBuildDirectory().file("module-dependencies.dot").get().getAsFile();
        outputFile.getParentFile().mkdirs();
        try (FileWriter writer = new FileWriter(outputFile)) {
            exporter.exportGraph(graph, writer);
        }
        
        project.getLogger().lifecycle("Module dependency graph generated at: " + outputFile.getAbsolutePath());
    }
} 

File: src/main/kotlin/com/sweetraingarden/gplugin/moduledot/ModuleDotPlugin.kt
package com.sweetraingarden.gplugin.moduledot

import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.artifacts.ProjectDependency
import org.jgrapht.Graph
import org.jgrapht.graph.DefaultDirectedGraph
import org.jgrapht.graph.DefaultEdge
import org.jgrapht.nio.DefaultAttribute
import org.jgrapht.nio.dot.DOTExporter
import java.io.FileWriter

class ModuleDotPlugin : Plugin<Project> {
    override fun apply(project: Project) {
        project.tasks.register("generateModuleDotGraph") { task ->
            task.group = "documentation"
            task.description = "Generates a DOT graph of module dependencies"
            
            task.doLast {
                try {
                    generateDotGraph(project)
                } catch (e: Exception) {
                    throw RuntimeException("Failed to generate module dependency graph", e)
                }
            }
        }
    }

    internal fun generateDotGraph(project: Project) {
        val graph: Graph<String, DefaultEdge> = DefaultDirectedGraph(DefaultEdge::class.java)
        
        // Build graph using project paths as unique identifiers
        buildProjectGraph(project, graph)
        
        // Create DOT exporter with custom vertex labeling
        val exporter = DOTExporter<String, DefaultEdge>()
        exporter.setVertexAttributeProvider { vertex ->
            mapOf(
                "label" to DefaultAttribute.createAttribute(getProjectNameFromPath(vertex)),
                "shape" to DefaultAttribute.createAttribute("box"),
                "tooltip" to DefaultAttribute.createAttribute(vertex)
            )
        }

        // Export to file
        val outputFile = project.layout.buildDirectory.file("module-dependencies.dot").get().asFile
        outputFile.parentFile.mkdirs()
        FileWriter(outputFile).use { writer ->
            exporter.exportGraph(graph, writer)
        }
        
        project.logger.lifecycle("Module dependency graph generated at: ${outputFile.absolutePath}")
    }
    
    private fun buildProjectGraph(rootProject: Project, graph: Graph<String, DefaultEdge>) {
        // First, create a map of project paths to their configurations
        val projectMap = mutableMapOf<String, Project>()
        
        // Collect all projects and their paths
        rootProject.allprojects.forEach { project ->
            projectMap[project.path] = project
            graph.addVertex(project.path)
        }
        
        // Process dependencies using the map
        projectMap.values.forEach { project ->
            project.configurations.forEach { config ->
                config.dependencies
                    .filterIsInstance<ProjectDependency>()
                    .map { it.path }  // Use path instead of dependencyProject
                    .distinct()
                    .forEach { dependencyPath ->
                        if (projectMap.containsKey(dependencyPath)) {
                            graph.addEdge(project.path, dependencyPath)
                        }
                    }
            }
        }
    }
    
    private fun getProjectNameFromPath(path: String): String {
        return path.substringAfterLast(':').ifEmpty { "root" }
    }
} 